File: sudoku.rb

class Position
    attr_reader :row
    attr_reader :col
    def initialize(row = 0, col = 0)
        @row = row
        @col = col
    end
end

class Board
    def initialize()
        @board = 9.times { }
    end
    def get(pos)
        return @board[pos.row][pos.col]
    end
    def place(pos, value)
        @board[pos.row][pos.col] = value
    end

    def [](pos)
        return self.get(pos)
    end

    def []=(pos, value)
        return self.place(pos, value)
    end
end

b = Board.new
pos = Position.new
b[pos] = 10
puts b[pos]

File: sudoku.py
from board import Board

def advance(pos):
    (row, col) = pos
    col += 1
    if col >= 9:
        col = 0
        row += 1
    next = (row, col)
    return next


def solve(board, pos = (0, 0)):
    if pos[0] >= 9:
        return True
    
    if board.get(pos) != 0:
        return solve(board, advance(pos))

    for attempt in range(1, 10):
        if board.valid_move(pos, attempt):
            board.place(pos, attempt)
            if solve(board, advance(pos)):
                return True
            else:
                board.place(pos, 0)
    return False

def input_board():
    board = []
    for i in range(9):
        line = input()
        row = []
        for str in line.split(" "):
            num = int(str)
            row.append(num)
        board.append(row)
    return board

board = Board(input_board())
solve(board)
print()
print(board)
